{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "61ce6afd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import packages\n",
    "import json\n",
    "import importlib.util\n",
    "import pandas as pd\n",
    "\n",
    "# import baltic\n",
    "baltic_path = '/Users/jort/coding/baltic/baltic/baltic.py' # path to baltic.py file (https://github.com/evogytis/baltic)\n",
    "def load_module(name, path):\n",
    "    module_spec = importlib.util.spec_from_file_location(name, path)\n",
    "    module = importlib.util.module_from_spec(module_spec)\n",
    "    module_spec.loader.exec_module(module)\n",
    "    return module\n",
    "bt = load_module('bt', baltic_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "id": "1b4b61f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define paths, excluded tips, and initializing mutations\n",
    "# update with argparser for final version\n",
    "\n",
    "# path to input json tree file\n",
    "# json_file = '/a/b/c.json'\n",
    "# will provide this as a command line arg\n",
    "json_file = '/Users/jort/coding/h5-clades/20231107-h5nx-base-dropunknown-augurclades-1647C/auspice/flu_avian_h5nx_ha.json'\n",
    "\n",
    "# path to output mutations tsv file\n",
    "# output_file = '/x/y/z.tsv'\n",
    "# will provide this as a command line arg\n",
    "output_file = '/Users/jort/Desktop/20240108_h5nx_muts.tsv',\n",
    "\n",
    "# excluded tips\n",
    "# will provide these as a .txt config file (one strain per line)\n",
    "excluded_tips = ['A/environment/Changsha/213/2014', 'A/duck/Bangladesh/17D1012/2018']\n",
    "\n",
    "# initializing mutations\n",
    "# these are any mutations that you would like to be forced into the tsv\n",
    "# e.g., to assign the clade of basal outgroup sequences as 'unassigned'\n",
    "# will provide these as a .txt config file (one mutation per line, space-delimited to 'clade gene position alt')\n",
    "init_muts = '''0 HA 17 D\n",
    "'''\n",
    "# init_muts = '''unassigned HA 1 M\n",
    "# 0 HA 17 D\n",
    "# '''\n",
    "\n",
    "# convert space-delimited text into a list of lists, with each sublist containing an single line split by spaces\n",
    "# then transpose it such that the resulting list has sublists containing each element of the output mutation tsv\n",
    "# i.e., [[clades], [genes], [positions], [alts], [uniques]] (where uniques is a temp column for filtering)\n",
    "init_muts = list(map(list, zip(*[line.split(' ')[:4]+[True] for line in init_muts.split('\\n') if line.count(' ') >= 3])))\n",
    "\n",
    "# if init_muts is a single empty list (no parseable mutations provided), convert into a list of five empty lists\n",
    "# this ensures that when the clade mutation lists are first created, an error isn't thrown\n",
    "if init_muts == []:\n",
    "    init_muts = [[]]*5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "id": "0ed453ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define functions\n",
    "\n",
    "def bt_read_in_tree_json(input_tree):\n",
    "    '''read in a tree in json format'''\n",
    "    with open(input_tree) as json_file:\n",
    "        json_tree = json.load(json_file)\n",
    "    json_translation = {'absoluteTime':lambda k: k.traits['node_attrs']['num_date']['value'],'name':'name'} ## allows baltic to find correct attributes in JSON, height and name are required at a minimum\n",
    "    bt_tree, meta = bt.loadJSON(json_tree, json_translation)\n",
    "    return json_tree, bt_tree\n",
    "\n",
    "def return_nt_muts_on_branch(branch):\n",
    "    '''return list of nucleotide mutations on a given branch'''\n",
    "    nt_muts = []\n",
    "    if 'branch_attrs' in branch.traits:\n",
    "        if 'mutations' in branch.traits['branch_attrs']:\n",
    "            if 'nuc' in branch.traits['branch_attrs']['mutations']:\n",
    "                nt_muts = branch.traits['branch_attrs']['mutations']['nuc']\n",
    "    return nt_muts\n",
    "\n",
    "def return_ha_muts_on_branch(branch):\n",
    "    '''return list of ha amino acid mutations on a given branch'''\n",
    "    ha_muts = []\n",
    "    if 'branch_attrs' in branch.traits:\n",
    "        if 'mutations' in branch.traits['branch_attrs']:\n",
    "            if 'HA' in branch.traits['branch_attrs']['mutations']:\n",
    "                ha_muts = branch.traits['branch_attrs']['mutations']['HA'] \n",
    "    return ha_muts\n",
    "\n",
    "def return_all_muts_between_lcas(starting_node, ending_node, muts = None, i = None):\n",
    "    muts = muts or {'nuc': [], 'ha': []}\n",
    "    i = i or 0\n",
    "    \n",
    "    # all leaves that descend from the ending node\n",
    "    # will use these to determine if the ending node descends from the node currently being analyzed\n",
    "    end_leaves = ending_node.leaves\n",
    "    \n",
    "    # set an empty list of mutations and enumerate the children of the starting node; children can be tips or nodes\n",
    "    children = starting_node.children\n",
    "    \n",
    "    for child in children:        \n",
    "        \"\"\"if the child is a leaf:\n",
    "            move on, too far down the tree (we want a node, not terminal tip)\"\"\"\n",
    "        \"\"\"if the child is an internal node:\n",
    "            first, test whether that child node contains the target tips in its children.\n",
    "            child.leaves will output a list of the names of all tips descending from that node. if not, pass. \n",
    "            if the node does contain the target end tip in its leaves, keep traversing down that node recursively, \n",
    "            collecting mutations as you go\"\"\"\n",
    "\n",
    "        if child.branchType == \"leaf\":\n",
    "            # if child is a leaf, we went too far\n",
    "            pass\n",
    "        \n",
    "        elif child.branchType == \"node\":\n",
    "            # if child is a node, check if it's the ending node\n",
    "            ## if it is, add the branch muts and return the list of muts\n",
    "            ## if it's not the ending node, check if the ending node is in its children\n",
    "            ### if so, add muts and keep going\n",
    "            ### if not, we are not on the path to the ending node. pass\n",
    "            if child == ending_node:\n",
    "                # found ending node\n",
    "                nt_muts = return_nt_muts_on_branch(child)\n",
    "                muts['nuc'].append(nt_muts)\n",
    "                ha_muts = return_ha_muts_on_branch(child)\n",
    "                muts['ha'].append(ha_muts)\n",
    "                \n",
    "            elif all(leaf in child.leaves for leaf in end_leaves):\n",
    "                # found node on path to ending node\n",
    "                nt_muts = return_nt_muts_on_branch(child)\n",
    "                muts['nuc'].append(nt_muts)\n",
    "                ha_muts = return_ha_muts_on_branch(child)\n",
    "                muts['ha'].append(ha_muts)\n",
    "                \n",
    "                # continue iterating down the path\n",
    "                return_all_muts_between_lcas(child, ending_node, muts, i)\n",
    "                \n",
    "            else:\n",
    "                # node not on correct path\n",
    "                pass\n",
    "    \n",
    "    # flatten the list so that you don't have nested lists\n",
    "    nt_flat_list = [item for sublist in muts['nuc'] for item in sublist]\n",
    "    ha_flat_list = [item for sublist in muts['ha'] for item in sublist]\n",
    "    flat_list = [nt_flat_list, ha_flat_list]\n",
    "    return flat_list\n",
    "\n",
    "\n",
    "def get_clade_inheritance(tree, clade_relationships = None, parent = None):\n",
    "    '''return a dictionary of clade parent-child relationships, stored as {child: parent}'''\n",
    "    \n",
    "    clade_relationships = clade_relationships or {}\n",
    "    if tree.branchType == 'node':\n",
    "        if tree in bt_lcas.values():\n",
    "            clade = [cl for cl in bt_lcas.keys() if bt_lcas[cl] == tree][0]\n",
    "            clade_relationships[clade] = parent\n",
    "            parent = clade\n",
    "        children = tree.children\n",
    "        for child in children:\n",
    "            get_clade_inheritance(child, clade_relationships, parent)\n",
    "    return clade_relationships\n",
    "\n",
    "\n",
    "# get parent-child clade relationships\n",
    "def make_tree_from_relationships(clade_relationships_dict):\n",
    "    '''return a tree (as nested dicts) of clade relationships'''\n",
    "    \n",
    "    # convert dictionary into a list of (child, parent) tuples\n",
    "    # exclude if parent is None to allow for rooting on the most basal clade rather than None\n",
    "    relationships_list = [(k, v) for k, v in clade_relationships_dict.items() if v]\n",
    "    \n",
    "    # convert list of tuples into a nested dict tree\n",
    "    roots = set()\n",
    "    mapping = {}\n",
    "    for child, parent in relationships_list:\n",
    "        childitem = mapping.get(child, None)\n",
    "        if childitem is None:\n",
    "            childitem = {}\n",
    "            mapping[child] = childitem\n",
    "        else:\n",
    "            roots.discard(child)\n",
    "        parentitem = mapping.get(parent, None)\n",
    "        if parentitem is None:\n",
    "            mapping[parent] = {child:childitem}\n",
    "            roots.add(parent)\n",
    "        else:\n",
    "            parentitem[child] = childitem\n",
    "    return {i:mapping[i] for i in sorted(roots)}\n",
    "            \n",
    "            \n",
    "def get_clade_heights(tree, clade_heights = None, i = None):\n",
    "    '''return a dictionary of clade \"heights\", where the most basal clade has a\n",
    "    height of 0, its child clades a height of 1, grandchild clades 2, etc.'''\n",
    "    \n",
    "    clade_heights = clade_heights or {}\n",
    "    i = i or 0\n",
    "    for root in tree:\n",
    "        if i in clade_heights:\n",
    "            clade_heights[i].append(root)\n",
    "        else:\n",
    "            clade_heights[i] = [root]\n",
    "        if tree[root] != {}:\n",
    "            get_clade_heights(tree[root], clade_heights, i+1)\n",
    "    return clade_heights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "id": "45a4edfb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Tree height: 22.797827\n",
      "Tree length: 607.371549\n",
      "annotations present\n",
      "\n",
      "Numbers of objects in tree: 760 (354 nodes and 406 leaves)\n",
      "\n",
      "excluding A/duck/Bangladesh/17D1012/2018\n",
      "Clade 2.3.2.1c has only one member and thus no LCA can be found\n"
     ]
    }
   ],
   "source": [
    "## get clade defining mutations for all clades, excluding -like clades\n",
    "\n",
    "## exclude any tips which appear to be mislabled (e.g., the tip clusters with a different clade)\n",
    "## doing so will ensure that the called LCA for the clade will be the LCA of all sequences which are clustering well\n",
    "\n",
    "## need to add in filter for polybasic cleavage site mutations\n",
    "    ## find the coordinates for both HA and nuc\n",
    "    ## if mutation in the coordinate range, do not add to list\n",
    "\n",
    "#json_file = '/Users/jort/coding/h5-clades/10-reps-h5nx-ref-all-clades-2321c-2344-readded/rep1/auspice/flu_avian_h5nx_ha.json'\n",
    "#json_file = '/Users/jort/coding/h5-clades/WHO-2321-outgroup/auspice/flu_avian_h5nx_ha.json'\n",
    "#json_file = '/Users/jort/coding/h5-clades/20231107-h5nx-base-dropunknown/auspice/flu_avian_h5nx_ha.json'\n",
    "json_file = '/Users/jort/coding/h5-clades/20231107-h5nx-base-dropunknown-augurclades-1647C/auspice/flu_avian_h5nx_ha.json'\n",
    "\n",
    "json_tree, bt_tree = bt_read_in_tree_json(json_file)\n",
    "\n",
    "# make dictionary with each clade as a key and a list of all leaves (as baltic objects) of that clade as values\n",
    "clade_bt_leaves = {}\n",
    "for leaf in bt_tree.getExternal():\n",
    "    name = leaf.traits['name']\n",
    "    if name in excluded_tips:\n",
    "        print('excluding', name)\n",
    "    elif 'clade' in leaf.traits:\n",
    "        clade = leaf.traits['clade']\n",
    "        if not clade in clade_bt_leaves:\n",
    "            clade_bt_leaves[clade]  = [leaf]\n",
    "        else:\n",
    "            clade_bt_leaves[clade].append(leaf)\n",
    "#         if '-like' not in clade:\n",
    "#             if not clade in clade_bt_leaves:\n",
    "#                 clade_bt_leaves[clade]  = [leaf]\n",
    "#             else:\n",
    "#                 clade_bt_leaves[clade].append(leaf)\n",
    "\n",
    "# find the branch that is the last common ancestor (lca) to all leaves in each clade\n",
    "# store it in a clade:lca dict\n",
    "bt_lcas = {}  \n",
    "for clade, leaf_list in clade_bt_leaves.items():\n",
    "    try:\n",
    "        ancestor = bt_tree.commonAncestor(leaf_list)\n",
    "        # sometimes baltic commonAncestor returns an 'empty' node when the true lca should be the root of the tree\n",
    "        # check if the ancestor is empty (traits is empty dict {}) — if so, assign the tree root as lca instead\n",
    "        if ancestor.traits != {}:\n",
    "            bt_lcas[clade] = ancestor\n",
    "        else:\n",
    "            bt_lcas[clade] = bt_tree.root\n",
    "    except AssertionError:\n",
    "        # baltic commonAncestor throws error if there are < 2 tips in list\n",
    "        # if so, can't find an lca — pull mutations from the branch instead\n",
    "        print('Clade', clade, 'has only one member and thus no LCA can be found')\n",
    "        bt_lcas[clade] = leaf_list[0]\n",
    "\n",
    "# pull the mutations from each lca branch\n",
    "# store them in a clade:mutations dict\n",
    "bt_lcas_mutations = {}\n",
    "for clade, lca in bt_lcas.items():\n",
    "    try:\n",
    "        bt_lcas_mutations[clade] = lca.traits['branch_attrs']['mutations']\n",
    "    except KeyError:\n",
    "        print('LCA of clade', clade, 'has no mutations on the branch and will not be included')\n",
    "        \n",
    "\n",
    "# determine parent-child clade relationships\n",
    "clade_relationships = get_clade_inheritance(bt_tree.root, {})\n",
    "#clade_relationships_list = [(k, v) for k, v in clade_relationships.items()]\n",
    "clade_relationships_tree = make_tree_from_relationships(clade_relationships)\n",
    "clade_heights = get_clade_heights(clade_relationships_tree)\n",
    "\n",
    "# dict of mutations that occur between the LCA of a parent clade and the LCA of a clade of interest\n",
    "# e.g., lca_to_lca_muts['2.3.2.1d'] = [[nucmut1, nucmut2, ...], [hamut1, hamut2, ...]]\n",
    "## where nucmut1, nucmut2, ... are all the nt mutations that occurred between 2.3.2.1c LCA and 2.3.2.1d LCA\n",
    "## and hamut1, hamut2, ... are the ha mutations that occurred between the LCAs\n",
    "lca_to_lca_muts = {}\n",
    "lca_to_lca_muts_positions = {}\n",
    "\n",
    "for i in [j for j in clade_heights if j > 0]:\n",
    "    for clade in clade_heights[i]:\n",
    "        parent = clade_relationships[clade]\n",
    "        if clade in bt_lcas and parent in bt_lcas:\n",
    "            lca1 = bt_lcas[parent]\n",
    "            lca2 = bt_lcas[clade]\n",
    "            lca_to_lca_muts[clade] = return_all_muts_between_lcas(lca1, lca2)\n",
    "            nt_muts = lca_to_lca_muts[clade][0]\n",
    "            ha_muts = lca_to_lca_muts[clade][1]\n",
    "            nt_muts_positions = [mut[1:-1] for mut in nt_muts]\n",
    "            ha_muts_positions = [mut[1:-1] for mut in ha_muts]\n",
    "            \n",
    "            # if site toggles multiple times between lcas, need to remove all but the final mutation\n",
    "            # for instance, if parent lca has nt 100A and child lca has 100T, but 100C is an intermediate,\n",
    "            # then we will see both '100C' and '100T' as lca-to-lca muts but we only want the final 100T\n",
    "            # thus, need to remove the first n-1 muts of n muts with identical position values\n",
    "            rep_nt_muts_positions = set(pos for pos in nt_muts_positions if nt_muts_positions.count(pos) > 1)\n",
    "            rep_ha_muts_positions = set(pos for pos in ha_muts_positions if ha_muts_positions.count(pos) > 1)\n",
    "\n",
    "            for nt_pos in rep_nt_muts_positions:\n",
    "                rep_nt_muts = [mut for mut in nt_muts if mut[1:-1] == nt_pos]\n",
    "                for mut in rep_nt_muts[:-1]:\n",
    "                    nt_muts.remove(mut)\n",
    "            for ha_pos in rep_ha_muts_positions:\n",
    "                rep_ha_muts = [mut for mut in ha_muts if mut[1:-1] == ha_pos]\n",
    "                for mut in rep_ha_muts[:-1]:\n",
    "                    ha_muts.remove(mut)\n",
    "            \n",
    "            for nt_mut in [nt_mut for nt_mut in nt_muts if nt_mut in bt_lcas_mutations[clade]['nuc']]:\n",
    "                nt_muts.remove(nt_mut)\n",
    "                \n",
    "            if 'HA' in bt_lcas_mutations[clade]:\n",
    "                for ha_mut in [ha_mut for ha_mut in ha_muts if ha_mut in bt_lcas_mutations[clade]['HA']]:\n",
    "                    ha_muts.remove(ha_mut)\n",
    "            \n",
    "            nt_muts_positions = [mut[1:-1] for mut in nt_muts]\n",
    "            ha_muts_positions = [mut[1:-1] for mut in ha_muts]\n",
    "            \n",
    "            lca_to_lca_muts_positions[clade] = [nt_muts_positions, ha_muts_positions]\n",
    "            \n",
    "\n",
    "# convert the clade:mutations dict to a df in the correct format for augur clades\n",
    "clade_data = ['clade'] + init_muts[0] # clade that is being defined\n",
    "muttype_data = ['gene'] + init_muts[1] # type of mutation (nuc or HA)\n",
    "mutsite_data = ['site'] + init_muts[2] # site of mutation (bp or aa position)\n",
    "mut_data = ['alt'] + init_muts[3] # alternative \n",
    "unique_data = ['unique'] + init_muts[4] # unique mutation (not propagated from parental clade)\n",
    "\n",
    "\n",
    "for i in [j for j in clade_heights if j > 0]:\n",
    "    for clade in clade_heights[i]:\n",
    "        # get mutations from dict and append data to lists as appropriate\n",
    "        mutations = bt_lcas_mutations[clade]        \n",
    "        if clade_relationships[clade] != None and clade_relationships[clade] != '0':\n",
    "            clade_data.append(clade)\n",
    "            muttype_data.append('clade')\n",
    "            mutsite_data.append(clade_relationships[clade])\n",
    "            mut_data.append('')\n",
    "            unique_data.append(True)\n",
    "        if 'nuc' in mutations:\n",
    "            for mutation in [mut for mut in mutations['nuc'] if 'N' not in mut and '-' not in mut]:\n",
    "                # if mutation site is between 1036 and 1059 (inclusive) it is part of polybasic cleavage site -- ignore\n",
    "                if int(mutation[1:-1]) not in range(1036, 1060):\n",
    "                    clade_data.append(clade)\n",
    "                    muttype_data.append('nuc')\n",
    "                    mutsite_data.append(mutation[1:-1])\n",
    "                    mut_data.append(mutation[-1])\n",
    "                    unique_data.append(True)\n",
    "        if 'HA' in mutations:\n",
    "            for mutation in [mut for mut in mutations['HA'] if 'X' not in mut and '-' not in mut]:\n",
    "                # if mutation site is between 339 and 346 (inclusive) it is part of polybasic cleavage site -- ignore\n",
    "                if int(mutation[1:-1]) not in range(339, 347):\n",
    "                    clade_data.append(clade)\n",
    "                    muttype_data.append('HA')\n",
    "                    mutsite_data.append(mutation[1:-1])\n",
    "                    mut_data.append(mutation[-1])\n",
    "                    unique_data.append(True)\n",
    "\n",
    "        # explicitly add mutations propagated from parent clade to lists\n",
    "        parent = clade_relationships[clade]\n",
    "        while parent != None:            \n",
    "            parent_mutations = bt_lcas_mutations[parent]\n",
    "\n",
    "            if 'nuc' in parent_mutations:\n",
    "                for mutation in [mut for mut in parent_mutations['nuc'] if 'N' not in mut and '-' not in mut]:\n",
    "                    # if mutation site is between 1036 and 1059 (inclusive) it is part of polybasic cleavage site -- ignore\n",
    "                    if int(mutation[1:-1]) not in range(1036, 1060):\n",
    "                        if mutation[1:-1] not in lca_to_lca_muts_positions[clade][0]:\n",
    "                            clade_data.append(clade)\n",
    "                            muttype_data.append('nuc')\n",
    "                            mutsite_data.append(mutation[1:-1])\n",
    "                            mut_data.append(mutation[-1])\n",
    "                            unique_data.append(False)\n",
    "                        else:\n",
    "                            corr_mutation = [mut for mut in lca_to_lca_muts[clade][0] if mut[1:-1]==mutation[1:-1]][0]\n",
    "                            clade_data.append(clade)\n",
    "                            muttype_data.append('nuc')\n",
    "                            mutsite_data.append(corr_mutation[1:-1])\n",
    "                            mut_data.append(corr_mutation[-1])\n",
    "                            unique_data.append(True)\n",
    "            if 'HA' in parent_mutations:\n",
    "                for mutation in [mut for mut in parent_mutations['HA'] if 'X' not in mut and '-' not in mut]:\n",
    "                    # if mutation site is between 339 and 346 (inclusive) it is part of polybasic cleavage site -- ignore\n",
    "                    if int(mutation[1:-1]) not in range(339, 347):\n",
    "                        if mutation[1:-1] not in lca_to_lca_muts_positions[clade][1]:\n",
    "                            clade_data.append(clade)\n",
    "                            muttype_data.append('HA')\n",
    "                            mutsite_data.append(mutation[1:-1])\n",
    "                            mut_data.append(mutation[-1])\n",
    "                            unique_data.append(False)\n",
    "                        else:\n",
    "                            corr_mutation = [mut for mut in lca_to_lca_muts[clade][1] if mut[1:-1]==mutation[1:-1]][0]\n",
    "                            clade_data.append(clade)\n",
    "                            muttype_data.append('HA')\n",
    "                            mutsite_data.append(corr_mutation[1:-1])\n",
    "                            mut_data.append(corr_mutation[-1])\n",
    "                            unique_data.append(True)\n",
    "            parent = clade_relationships[parent]\n",
    "\n",
    "df = pd.DataFrame(list(zip(clade_data, muttype_data, mutsite_data, mut_data, unique_data)))\n",
    "df = df[(df[4]==True) | (df[4]=='unique')].drop(columns=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "id": "65e48b7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>clade</td>\n",
       "      <td>gene</td>\n",
       "      <td>site</td>\n",
       "      <td>alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>HA</td>\n",
       "      <td>17</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>nuc</td>\n",
       "      <td>223</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>nuc</td>\n",
       "      <td>519</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>nuc</td>\n",
       "      <td>606</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1809</th>\n",
       "      <td>2.2.1.1a</td>\n",
       "      <td>HA</td>\n",
       "      <td>87</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1810</th>\n",
       "      <td>2.2.1.1a</td>\n",
       "      <td>HA</td>\n",
       "      <td>178</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1811</th>\n",
       "      <td>2.2.1.1a</td>\n",
       "      <td>HA</td>\n",
       "      <td>201</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1812</th>\n",
       "      <td>2.2.1.1a</td>\n",
       "      <td>HA</td>\n",
       "      <td>291</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1813</th>\n",
       "      <td>2.2.1.1a</td>\n",
       "      <td>HA</td>\n",
       "      <td>421</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>784 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             0     1     2    3\n",
       "0        clade  gene  site  alt\n",
       "1            0    HA    17    D\n",
       "2            3   nuc   223    C\n",
       "3            3   nuc   519    C\n",
       "4            3   nuc   606    T\n",
       "...        ...   ...   ...  ...\n",
       "1809  2.2.1.1a    HA    87    P\n",
       "1810  2.2.1.1a    HA   178    E\n",
       "1811  2.2.1.1a    HA   201    E\n",
       "1812  2.2.1.1a    HA   291    S\n",
       "1813  2.2.1.1a    HA   421    K\n",
       "\n",
       "[784 rows x 4 columns]"
      ]
     },
     "execution_count": 302,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# view clade-defining mutations dataframe\n",
    "# note that the initializing mutation (0 HA 17 D) is the first entry\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "78f5c888",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save mutations to tsv\n",
    "\n",
    "# output_file = '/x/y/z.tsv'\n",
    "# df.to_csv(output_file, sep=\"\\t\", index=False, header=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
